### PR TITLE
Add a scalar replacement pass

### PR DESCRIPTION
**Pull request by alan-baker**
The pass considers all function scopevariables of composite type. If there are accesses to individual elements (and it is legal) the pass replaces the variable with a variable for each composite element and updates all the uses.

Added the pass to -O (not -Os)
Added NumUses and NumUsers to DefUseManager
Added some helper methods for the inst to block mapping in context
Added some helper methods for specific constant types

### CODE REVIEW

File: source/opt/scalar_replacement_pass.cpp [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+  CheckUses(inst, &stats, &ok);
+
+  // TODO(alanbaker): Extend this to some meaningful heuristics about when SRoA
+  // is valuable.
+  if (stats.num_partial_accesses == 0) ok = false;
+
+  return ok;
+}
+
+void ScalarReplacementPass::CheckUses(const ir::Instruction* inst,
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Why not have it return a bool instead of having |ok| as a parameter?

File: source/opt/scalar_replacement_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+        switch (user->opcode()) {
+          case SpvOpAccessChain:
+          case SpvOpInBoundsAccessChain:
+            if (index == 2u) {
+              uint32_t id = user->GetSingleWordOperand(3u);
+              const ir::Instruction* opInst = get_def_use_mgr()->GetDef(id);
+              if (!ir::IsCompileTimeConstantInst(opInst->opcode())) {
+                *ok = false;
+              } else {
+                CheckUsesRelaxed(user, ok);
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Not obvious  what you are looking for and why.  Can you add to the comments in the header file?  My guess is that you want to make sure the address does not "escape".  However, why would if matter if the address of the member escapes as say a parameter on a function call?  You can pass it the address of the new variable.

File: source/opt/scalar_replacement_pass.cpp [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+}
+
+bool ScalarReplacementPass::CheckUses(const ir::Instruction* inst) const {
+  bool ok = true;
+  VariableStats stats = {0, 0};
+  CheckUses(inst, &stats, &ok);
+
+  // TODO(alanbaker): Extend this to some meaningful heuristics about when SRoA
+  // is valuable.
+  if (stats.num_partial_accesses == 0) ok = false;
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: +1

File: source/opt/scalar_replacement_pass.cpp [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+bool ScalarReplacementPass::ReplaceVariable(
+    ir::Instruction* inst, std::queue<ir::Instruction*>* worklist) {
+  std::vector<ir::Instruction*> replacements;
+  CreateReplacementVariables(inst, &replacements);
+
+  bool ok = true;
+  std::vector<ir::Instruction*> dead;
+  dead.push_back(inst);
+  get_def_use_mgr()->ForEachUser(
+      inst, [this, &ok, replacements, &dead](ir::Instruction* user) {
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Looks like a value capture of |replacements|, could this cause many call to the copy constructor for the vector?

File: source/opt/scalar_replacement_pass.cpp [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+      constantId = iter->second;
+    }
+
+    // Create a new constant extract.
+    newInitId = TakeNextId();
+    context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+        context(), SpvOpSpecConstantOp, storageId, newInitId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER, {SpvOpCompositeExtract}},
+            {SPV_OPERAND_TYPE_ID, {constantId}}}));
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Are you sure this is the correct instruction?  An OpCompositeExtract takes more than 1 operand.

File: source/opt/scalar_replacement_pass.cpp [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+  } else if (init->opcode() == SpvOpConstantComposite) {
+    // Get the appropriate index constant.
+    newInitId = init->GetSingleWordInOperand(index);
+    ir::Instruction* element = get_def_use_mgr()->GetDef(newInitId);
+    if (element->opcode() == SpvOpUndef) {
+      // Undef is not a valid initializer for a variable.
+      newInitId = 0;
+    }
+  } else {
+    assert(false);
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: The spec specifically states that the initializer could be the id of a global variable.  Add a comment that you are specifically disallowing it.  Also I don't see the code in |CanReplaceVariable| that disallows this case.  Should be added there.

File: source/opt/scalar_replacement_pass.cpp [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+  for (uint32_t i = 0; i != op.words.size(); ++i) {
+    len |= (op.words[i] << (32 * i));
+  }
+  return len;
+}
+
+size_t ScalarReplacementPass::GetConstantInteger(
+    const ir::Instruction* constant) const {
+  assert(constant->opcode() == SpvOpConstant ||
+         constant->opcode() == SpvOpConstantNull);
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Should assert in the type as well?  OpConstant is uses for floating point constants as well.

File: source/opt/scalar_replacement_pass.cpp [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+
+  uint32_t ptrTypeId = inst->type_id();
+  uint32_t typeId =
+      get_def_use_mgr()->GetDef(ptrTypeId)->GetSingleWordInOperand(1u);
+  return get_def_use_mgr()->GetDef(typeId);
+}
+
+bool ScalarReplacementPass::CanReplaceVariable(
+    const ir::Instruction* varInst) const {
+  assert(varInst->opcode() == SpvOpVariable);
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Might want to check for decorations on variables as well.  A variable could be decorated as volatile for example.

File: source/opt/scalar_replacement_pass.cpp [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+    default:
+      assert(false && "Unexpected type.");
+      break;
+  }
+}
+
+void ScalarReplacementPass::CreateVariable(
+    uint32_t typeId, ir::Instruction* varInst, uint32_t index,
+    std::vector<ir::Instruction*>* replacements) {
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: Somewhere in here you might need to copy decorations from the original struct to the new variable.

File: source/opt/scalar_replacement_pass.cpp [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+      if (GetArrayLength(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+    // Specifically including matrix and vector in an attempt to reduce the
+    // number of vector registers required.
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      if (GetNumElements(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+
+    case SpvOpTypeRuntimeArray:
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I just wanted to explicitly call out that it should not be handled despite being an aggregate type.

File: source/opt/scalar_replacement_pass.cpp (hunk 1 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Convenience due to the recursive nature and interaction with ForEachUser lambda.

File: source/opt/scalar_replacement_pass.cpp (hunk 2 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: This checks that the pointer ends up in a load or store and not something else. If this were a kernel, I could do ptrtoint and then weird stuff. I don't want that to be allowed.

File: source/opt/scalar_replacement_pass.cpp (hunk 9 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I'm not sure that's appropriate, but I will peruse the decorations section in the spec.

File: source/opt/scalar_replacement_pass.cpp (hunk 8 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Function scope variables cannot be decorated as volatile, but I'll see if there are other decorations that are inappropriate.

File: source/opt/scalar_replacement_pass.cpp (hunk 7 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Sure I can add that.

File: source/opt/scalar_replacement_pass.cpp (hunk 6 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Since OpVariable returns a pointer, I won't see any valid cases for scalarization (only composite types are handled).

File: source/opt/scalar_replacement_pass.cpp (hunk 5 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I'll double check. It passes the validator though.

File: source/opt/scalar_replacement_pass.cpp (hunk 4 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Good catch.

File: source/opt/ir_context.h [hunk 11]
```diff
@@ -192,6 +192,20 @@ class IRContext {
     auto entry = instr_to_block_.find(instr);
     return (entry != instr_to_block_.end()) ? entry->second : nullptr;
   }
+  ir::BasicBlock* get_instr_block(uint32_t id) {
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/ir_context.h]: Ok

File: source/opt/scalar_replacement_pass.cpp (hunk 5 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: You're right, I'll fix this.

File: source/opt/scalar_replacement_pass.cpp (hunk 9 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I've added code to handle decorations.

File: source/opt/scalar_replacement_pass.cpp [hunk 12]
```diff
@@ -0,0 +1,702 @@
...
+              assert(false && "Unexpected opcode");
+              break;
+          }
+        }
+      });
+
+  // There was an illegal access.
+  if (!ok) return false;
+
+  // Clean up some dead code.
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Why not just leave this to DCE?  Makes the pass slightly cleaner.

File: source/opt/scalar_replacement_pass.cpp [hunk 13]
```diff
@@ -0,0 +1,702 @@
...
+    context()->set_instr_block(&*iter, block);
+
+    // Create the store.
+    std::unique_ptr<ir::Instruction> newStore(
+        new ir::Instruction(context(), SpvOpStore, 0, 0,
+                            std::initializer_list<ir::Operand>{
+                                {SPV_OPERAND_TYPE_ID, {var->result_id()}},
+                                {SPV_OPERAND_TYPE_ID, {extractId}}}));
+    // Copy memory access attributes.
+    for (uint32_t i = 2; i < store->NumInOperands(); ++i) {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Why skipping the first two?  This needs a comment.

File: source/opt/scalar_replacement_pass.cpp [hunk 14]
```diff
@@ -0,0 +1,702 @@
...
+bool ScalarReplacementPass::ReplaceAccessChain(
+    ir::Instruction* chain, const std::vector<ir::Instruction*>& replacements) {
+  // Replaces the access chain with either another access chain (with one fewer
+  // indexes) or a direct use of the replacement variable.
+  uint32_t indexId = chain->GetSingleWordInOperand(1u);
+  const ir::Instruction* index = get_def_use_mgr()->GetDef(indexId);
+  size_t indexValue = GetConstantInteger(index);
+  if (indexValue > replacements.size()) {
+    // Out of bounds access, this is illegal IR.
+    return false;
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Easier to just assert?  Nothing good will happen from this point on.

File: source/opt/scalar_replacement_pass.cpp [hunk 15]
```diff
@@ -0,0 +1,702 @@
...
+    if (chain->NumInOperands() > 2) {
+      // Replace input access chain with another access chain.
+      ir::BasicBlock::iterator chainIter(chain);
+      uint32_t replacementId = TakeNextId();
+      std::unique_ptr<ir::Instruction> replacementChain(new ir::Instruction(
+          context(), chain->opcode(), chain->type_id(), replacementId,
+          std::initializer_list<ir::Operand>{
+              {SPV_OPERAND_TYPE_ID, {var->result_id()}}}));
+      // Add the remaining indexes.
+      for (uint32_t i = 2; i < chain->NumInOperands(); ++i) {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Likewise.  You're using *InOperands() so it's not immediately obvious why the loop starts at 2.

File: source/opt/scalar_replacement_pass.h [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+  // If there is an initial value for |source| for element |index|, it is
+  // appended as an operand on |newVar|. If the initial value is OpUndef, no
+  // initial value is added to |newVar|.
+  void GetOrCreateInitialValue(ir::Instruction* source, uint32_t index,
+                               ir::Instruction* newVar);
+
+  // Returns the result id of a 32 bit integer.
+  //
+  // Searches all types for a decoration-less integer before creating one.
+  uint32_t GetIntId();
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.h]: Why not use TypeManager?

File: source/opt/scalar_replacement_pass.cpp [hunk 17]
```diff
@@ -0,0 +1,702 @@
...
+
+  // If varInst was initialized, make sure to initialize its replacement.
+  GetOrCreateInitialValue(varInst, index, inst);
+  get_def_use_mgr()->AnalyzeInstDefUse(inst);
+  context()->set_instr_block(inst, block);
+
+  replacements->push_back(inst);
+}
+
+uint32_t ScalarReplacementPass::GetOrCreatePointerType(uint32_t id) {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: This is likely better implemented in analysis::TypeManager.

File: source/opt/scalar_replacement_pass.cpp [hunk 18]
```diff
@@ -0,0 +1,702 @@
...
+          {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassFunction}},
+          {SPV_OPERAND_TYPE_ID, {id}}}));
+  ir::Instruction* ptr = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDefUse(ptr);
+  pointee_to_pointer_[id] = ptrId;
+
+  return ptrId;
+}
+
+uint32_t ScalarReplacementPass::GetIntId() {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Likewise with this one.  The TypeManager would give you an existing Int, if it finds one (or should ;)

File: source/opt/scalar_replacement_pass.cpp [hunk 19]
```diff
@@ -0,0 +1,702 @@
...
+                                                    uint32_t index,
+                                                    ir::Instruction* newVar) {
+  assert(source->opcode() == SpvOpVariable);
+  if (source->NumInOperands() < 2) return;
+
+  uint32_t initId = source->GetSingleWordInOperand(1u);
+  uint32_t storageId = GetStorageType(newVar)->result_id();
+  ir::Instruction* init = get_def_use_mgr()->GetDef(initId);
+  uint32_t newInitId = 0;
+  if (init->opcode() == SpvOpConstantNull) {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: There is code in fold_spec_constant_op_and_composite_pass.cpp that does something similar, but I'm currently in the midst of re-factoring it for use with CCP.  Add a TODO for me to deal with this after the dust settles?

File: source/opt/scalar_replacement_pass.cpp [hunk 20]
```diff
@@ -0,0 +1,702 @@
...
+  } else {
+    assert(false);
+  }
+
+  if (newInitId != 0) {
+    newVar->AddOperand({SPV_OPERAND_TYPE_ID, {newInitId}});
+  }
+}
+
+size_t ScalarReplacementPass::GetIntegerLiteral(const ir::Operand& op) const {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: This looks like a useful method for Operand or Instruction.

File: source/opt/scalar_replacement_pass.cpp [hunk 21]
```diff
@@ -0,0 +1,702 @@
...
+  assert(op.words.size() <= 2);
+  size_t len = 0;
+  for (uint32_t i = 0; i != op.words.size(); ++i) {
+    len |= (op.words[i] << (32 * i));
+  }
+  return len;
+}
+
+size_t ScalarReplacementPass::GetConstantInteger(
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Likewise.  There's actually a few more.  Not sure if they all make sense, but any methods that are now taking an ir::Instruction* as parameter may be good for just adding inside ir::Instruction.

File: source/opt/scalar_replacement_pass.cpp [hunk 22]
```diff
@@ -0,0 +1,702 @@
...
+  if (!CheckTypeAnnotations(get_def_use_mgr()->GetDef(varInst->type_id())))
+    return false;
+
+  const ir::Instruction* typeInst = GetStorageType(varInst);
+  if (!CheckType(typeInst)) return false;
+
+  if (!CheckAnnotations(varInst)) return false;
+
+  if (!CheckUses(varInst)) return false;
+
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Consider just 'return CheckType && CheckAnnotations && CheckUses'?

File: source/opt/scalar_replacement_pass.cpp [hunk 23]
```diff
@@ -0,0 +1,702 @@
...
+      if (GetNumElements(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+
+    case SpvOpTypeRuntimeArray:
+    default:
+      return false;
+  }
+}
+
+bool ScalarReplacementPass::CheckTypeAnnotations(
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: The name of this predicate is confusing.  What is it checking?

File: source/opt/scalar_replacement_pass.cpp [hunk 24]
```diff
@@ -0,0 +1,702 @@
...
+        break;
+      default:
+        return false;
+    }
+  }
+
+  return true;
+}
+
+bool ScalarReplacementPass::CheckAnnotations(
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Likewise here.

File: source/opt/scalar_replacement_pass.cpp (hunk 1 above)
Comment (dnovillo) [reply source/opt/scalar_replacement_pass.cpp]: But you can capture &ok in the lambda and return it.

File: test/opt/pass_fixture.h [hunk 25]
```diff
@@ -89,11 +89,23 @@ class PassTest : public TestT {
   // disassembly string and the boolean value from the pass Process() function.
   template <typename PassT, typename... Args>
   std::tuple<std::string, opt::Pass::Status> SinglePassRunAndDisassemble(
-      const std::string& assembly, bool skip_nop, Args&&... args) {
+      const std::string& assembly, bool skip_nop, bool do_validation,
```
Comment (dnovillo) [inline test/opt/pass_fixture.h]: default do_validation to 'true'?

File: tools/opt/opt.cpp [hunk 26]
```diff
@@ -197,17 +195,23 @@ Options (in lexicographical order):
                'spirv-opt --merge-blocks -O ...' applies the transformation
                --merge-blocks followed by all the transformations implied by
                -O.
-  --relax-store-struct
+  --relax-struct-store
```
Comment (dnovillo) [inline tools/opt/opt.cpp]: I'd rather not change flag names here.  There may be external users relying on the previous name.  Better leave it for another PR.

File: source/opt/ir_context.h [hunk 27]
```diff
@@ -357,6 +384,12 @@ class IRContext {
     valid_analyses_ = valid_analyses_ | kAnalysisCFG;
   }
 
+  // Analyzes the features in the owned module. Builds the manager if required.
+  void AnalyzeFeatures() {
+    feature_mgr_.reset(new opt::FeatureManager());
+    feature_mgr_->Analyze(module());
+  }
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/ir_context.h]: I don't think we're planning on adding any optimizations that enable features, so it shouldn't change.

File: source/opt/scalar_replacement_pass.cpp (hunk 12 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: It makes it easier to test and I know exactly what is dead as opposed to scanning the whole module.

File: source/opt/scalar_replacement_pass.cpp (hunk 13 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Ok

File: source/opt/scalar_replacement_pass.cpp (hunk 14 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: This causes the pass to return Failure, which I think is better than only catching it in debug.

File: source/opt/scalar_replacement_pass.cpp (hunk 15 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: ok

File: source/opt/scalar_replacement_pass.cpp (hunk 17 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: The type manager is pretty bare bones right now. I agree this method would be better implemented elsewhere, but I'd rather have a separate change for that.

File: source/opt/scalar_replacement_pass.cpp (hunk 18 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: Same as above, I'd prefer to leave this refactor to another PR.

File: source/opt/scalar_replacement_pass.cpp (hunk 19 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: ok

File: source/opt/scalar_replacement_pass.cpp (hunk 23 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: As mentioned in the header it checks annotations on a type. This is different than CheckAnnotations because different annotations are acceptable for variables and types.

File: test/opt/pass_fixture.h (hunk 25 above)
Comment (alan-baker) [reply test/opt/pass_fixture.h]: I didn't want to make that choice for other tests. Its possible they are using invalid IR for their testcases.

File: tools/opt/opt.cpp (hunk 26 above)
Comment (alan-baker) [reply tools/opt/opt.cpp]: This option exists in the validator as --relax-struct-store, but in opt it was --relax-store-struct. This simply unifies those options since they do the same thing.

File: source/opt/scalar_replacement_pass.cpp (hunk 17 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I take it back, I'll make what changes I can to include the type manager.

File: source/opt/scalar_replacement_pass.cpp (hunk 18 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: This dead, I've removed it.

File: source/opt/scalar_replacement_pass.cpp (hunk 17 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I'll leave a todo to improve the TypeManager, its not very useful as is.

File: source/opt/scalar_replacement_pass.cpp (hunk 20 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.cpp]: I'd like to leave that for a larger discussion of sub-classes. Same with the comments about instruction functions.

File: source/opt/scalar_replacement_pass.h (hunk 16 above)
Comment (alan-baker) [reply source/opt/scalar_replacement_pass.h]: This is dead, removed.

File: source/opt/scalar_replacement_pass.cpp [hunk 28]
```diff
@@ -0,0 +1,678 @@
...
+namespace opt {
+
+// Heuristic aggregate element limit.
+const uint32_t MAX_NUM_ELEMENTS = 100u;
+
+Pass::Status ScalarReplacementPass::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+
+  Status status = Status::SuccessWithoutChange;
+  for (auto& f : *get_module()) {
```
Comment (dnovillo) [inline source/opt/scalar_replacement_pass.cpp]: Use ProcessReachableCallTree() here?

File: source/opt/scalar_replacement_pass.cpp (hunk 12 above)
Comment (dnovillo) [reply source/opt/scalar_replacement_pass.cpp]: Hmm, OK.  I generally prefer to keep transformations simple, but I agree that this may not be a big deal.

File: source/opt/scalar_replacement_pass.cpp (hunk 28 above)
Comment (s-perron) [reply source/opt/scalar_replacement_pass.cpp]: Now that we have dead function elimination running, I don't think it makes a difference.

File: source/opt/scalar_replacement_pass.cpp [hunk 29]
```diff
@@ -0,0 +1,678 @@
...
+      return true;
+    case SpvOpTypeArray:
+      if (GetArrayLength(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+    // Specifically including matrix and vector in an attempt to reduce the
+    // number of vector registers required.
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      if (GetNumElements(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
```
Comment (s-perron) [inline source/opt/scalar_replacement_pass.cpp]: I looked at how this pass affects code.  I think for now it would be a good idea to not split matrices and vectors.  In another PR, we could see how to improve the heuristics for them.

**Comment by s-perron** [comment]
Rebased, squashed and merged as 867451f4.
